### PR TITLE
fix: honor request options

### PR DESCRIPTION
**Pull request by leonsomed**
## Description
This PR enforces honoring on queryParams, pathParams, and header like it was supported in version < 2.1.1.

## Motivation and Context
Previous to version 2.1.1 queryParams, pathParams, and header were honored to remain as the options that were explicitly passed to the request method. However, version 2.1.1 introduces a change where it will initialize queryParams, pathParams, and header to an empty object when they are not passed to the request method.

The behavior described above causes an issue with requests that contain queryParams in the path. For example: `/users?lucky=true`, this path contains an embeded queryParam but it will be stripped away before the execution because it is not included in the queryParams object. Prior to version 2.1.1 this was not a problem because an undefined queryParams will be ignored, however after version 2.1.1 an undefined queryParams will be initialized to an empty object and it will cause the removal of the embedded `lucky` query param.

The reason this is important is because there are APIs that respond with link resources with embedded queryParams.

## How Has This Been Tested?
Previously this call will strip away the lucky query param
`client.request({ method: 'GET', path: '/users?lucky=true' });`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by leonsomed** [comment]
The only affected option is `queryParams` both `pathParams` and `header` do not get affected.
After further analysis I realized queryParams were not being stripped but instead they were being obscured by an additional query string, ie:

```
// this requests the following path, notice the double starty query string character: ?

client.request({ method: 'GET', path: '/users?lucky=true', queryParams: {} });
// /users?lucky=true?

client.request({ method: 'GET', path: '/users?lucky=true', queryParams: { test: true } });
// /users?lucky=true?test=true
```

So the new approach ensures not breaking the existing query string in path, if it exists, ie:

```
client.request({ method: 'GET', path: '/users?lucky=true', queryParams: {} });
// /users?lucky=true&

client.request({ method: 'GET', path: '/users?lucky=true', queryParams: { test: true } });
// /users?lucky=true&test=true
```

File: lib/http/request.js [hunk 1]
```diff
@@ -190,7 +190,7 @@ const create = async function (method, path, options, hooks = {}) {
   })
 
   if (options.queryParams) {
-    path += `?${QueryString.stringify(options.queryParams)}`
+    path += `${path.includes('?') ? '&' : '?'}${QueryString.stringify(options.queryParams)}`
```
Comment (tuckbick) [inline lib/http/request.js:193]: I think we also need to account for if `path` already ends in a `&`. We wouldn't want to have `?foo=bar&&fizz=buzz`.

Some options to address this:
* strip any trailing `&` before processing anything
* don't add a `&` if one already exists
* parse the path's query string into an object, merge `options.queryParams` into the object, then stringify and re-attach to the path

Comment (leonsomed) [reply lib/http/request.js]: i took the third approach

**Comment by tuckbick** [comment]
@leonsomed Released in v2.1.2 and v3.0.1.

**Comment by programax** [comment]
thanks a lot!

**Pull request merged**